src/Reflection/FieldItemListPropertyReflection.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace PHPStan\Reflection;
+
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+/**
+ * Allows field access via magic methods
+ *
+ * See \Drupal\Core\Field\FieldItemListInterface::__get and ::__set.
+ */
+class FieldItemListPropertyReflection implements PropertyReflection
+{
+
+  /** @var ClassReflection */
+    private $declaringClass;
+
+  /** @var string */
+    private $propertyName;
+
+    public function __construct(ClassReflection $declaringClass, string $propertyName)
+    {
+        $this->declaringClass = $declaringClass;
+        $this->propertyName = $propertyName;
+    }
+
+    public static function canHandleProperty(string $propertyName): bool
+    {
+        $names = ['entity', 'value', 'target_id'];
+        return in_array($propertyName, $names, true);
+    }
+
+    public function getType(): Type
+    {
+        if ($this->propertyName == 'entity') {
+          // It was a EntityReferenceFieldItemList
+            return new ObjectType('Drupal\Core\Entity\FieldableEntityInterface');
+        }
+        if ($this->propertyName == 'target_id') {
+          // It was a EntityReferenceFieldItemList
+            return new IntegerType();
+        }
+        if ($this->propertyName == 'value') {
+            return new MixedType();
```

Comment:
Add a todo to provide some kind of resolving for this to provide a proper return type based on the assume field class.

Reply:
Afaik it's always a string when you load the entity from the Database. Even booleans are just '0' and '1'

Reply:
Yeah, it's always a string unless you do `getValue` on the property itself and TypedData casts it.